tests/Aql/NoResultsExecutorTest.cpp

```diff
@@ -0,0 +1,148 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2018 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Jan Christoph Uhde
+////////////////////////////////////////////////////////////////////////////////
+
+#include "BlockFetcherHelper.h"
+#include "catch.hpp"
+
+#include "Aql/AqlItemBlock.h"
+#include "Aql/AqlItemBlockShell.h"
+#include "Aql/ExecutionBlockImpl.h"
+#include "Aql/ExecutorInfos.h"
+#include "Aql/InputAqlItemRow.h"
+#include "Aql/NoResultsExecutor.h"
+#include "Aql/ResourceUsage.h"
+#include "Aql/SingleRowFetcher.h"
+
+#include <velocypack/Builder.h>
+#include <velocypack/velocypack-aliases.h>
+
+using namespace arangodb;
+using namespace arangodb::aql;
+
+namespace arangodb {
+namespace tests {
+namespace aql {
+
+SCENARIO("NoResultsExecutor", "[AQL][EXECUTOR][NORESULTS]") {
+  ExecutionState state;
+
+  ResourceMonitor monitor;
+  AqlItemBlockManager itemBlockManager(&monitor);
+  auto block = std::make_unique<AqlItemBlock>(&monitor, 1000, 1);
+  auto outputRegisters = make_shared_unordered_set({0});
+  auto registersToKeep = make_shared_unordered_set();
+  auto outputBlockShell =
+      std::make_unique<OutputAqlItemBlockShell>(itemBlockManager, std::move(block),
+                                                outputRegisters, registersToKeep);
+
+  REQUIRE(outputRegisters->size() == 1);
+  REQUIRE((*(outputRegisters->begin()) == 0));
+
+  RegisterId inputRegister(0);
+  RegisterId outputRegister(*(outputRegisters->begin()));
+
+  ExecutorInfos infos(make_shared_unordered_set({0}), make_shared_unordered_set(),
+                      1 /*nr in*/, 1 /*nr out*/, std::unordered_set<RegisterId>{});
+  OutputAqlItemRow result(std::move(outputBlockShell));
+
+  GIVEN("there are no rows upstream") {
+    VPackBuilder input;
+
+    WHEN("the producer does not wait") {
+      SingleRowFetcherHelper fetcher(input.steal(), false);
+      NoResultsExecutor testee(fetcher, infos);
+      NoStats stats{};
+
+      THEN("the executor should return DONE with nullptr") {
+        std::tie(state, stats) = testee.produceRow(result);
+        REQUIRE(state == ExecutionState::DONE);
+        REQUIRE(!result.produced());
```

Comment:
Here and below, the test message is somewhat misleading. It originally referred to `result.stealBlock() == nullptr`, but this isn't tested here (actually, we have this in other blocks already and should fix it there, too). Or add this test, which won't harm, either.